ReactNativeClient/lib/file-api-driver-amazon-s3.js

```diff
@@ -0,0 +1,327 @@
+		});
+	}
+
+	async s3UploadFileFrom(path, key) {
+		const body = await shim.readFileToBlob(path);
+
+		return new Promise((resolve, reject) => {
+			this.api().upload({
+				Bucket: this.s3_bucket_,
+				Key: key,
+				Body: body,
+			}, (err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	async s3DeleteObject(key) {
+		return new Promise((resolve, reject) => {
+			this.api().deleteObject({
+				Bucket: this.s3_bucket_,
+				Key: key,
+			},
+			(err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	// Assumes key is formatted, like `{Key: 's3 path'}`
+	async s3DeleteObjects(keys) {
+		return new Promise((resolve, reject) => {
+			this.api().deleteObjects({
+				Bucket: this.s3_bucket_,
+				Delete: { Objects: keys },
+			},
+			(err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	async stat(path) {
+		try {
+			const metadata = await this.s3HeadObject(this.makePath_(path));
+
+			return this.metadataToStat_(metadata, path);
+		} catch (error) {
+			if (this.hasErrorCode_(error, 'NotFound')) {
+				// ignore
+			} else {
+				throw error;
+			}
+		}
+	}
+
+	metadataToStat_(md, path) {
+		const output = {
+			path: path,
+			updated_time: md['LastModified'] ? new Date(md['LastModified']) : new Date(),
+			isDeleted: !!md['DeleteMarker'],
+			isDir: false,
+		};
+
+		return output;
+	}
+
+	metadataToStats_(mds) {
+		const output = [];
+		for (let i = 0; i < mds.length; i++) {
+			output.push(this.metadataToStat_(mds[i], mds[i].Key));
+		}
+		return output;
+	}
+
+	async setTimestamp() {
+		throw new Error('Not implemented'); // Not needed anymore
+	}
+
+	async delta(path, options) {
+		const getDirStats = async path => {
+			const result = await this.list(path);
+			return result.items;
+		};
+
+		return await basicDelta(path, getDirStats, options);
+	}
+
+	async list(path) {
+		let response = await this.s3ListObjects(this.makePath_(path));
+
+		let output = this.metadataToStats_(response.Contents);
+
+		while (response.IsTruncated) {
+			response = await this.s3ListObjects(this.makePath_(path), response.NextContinuationToken);
+
+			output = output.concat(this.metadataToStats_(response.Contents));
+		}
+
+		return {
+			items: output,
+			hasMore: false,
+			context: { cursor: response.NextContinuationToken },
+		};
+	}
+
+	async get(path, options) {
+		const remotePath = this.makePath_(path);
+		if (!options) options = {};
+		const responseFormat = options.responseFormat || 'text';
+
+		try {
+			let output = null;
+			const response = await this.s3GetObject(remotePath);
+			output = response.Body;
+
+
+			if (responseFormat === 'text') {
+				output = output.toString();
+			}
+
+			if (options.target === 'file') {
+				if (!options.path) throw new Error('get: target options.path is missing');
+
+				// TODO: check if this ever hits on RN
+				await shim.fsDriver().writeBinaryFile(options.path, output);
```

Comment:
@laurent22 I have a question here for React Native, how should the driver handle `get` for files on RN? `writeBinaryFile` will raise an exception on the RN shim.